### PR TITLE
Add tests for FIP-0019

### PR DESCRIPTION
**Pull request by arajasek**

### CODE REVIEW

**Comment by arajasek** [comment]
Rebased onto master

**Comment by arajasek** [comment]
Rebased

**Review by ZenGround0 — APPROVED** [review]
I left a decent amount of feedback but mostly looking good.  More tests would be good (as always) but this is a solid foundation.

The unwritten test I want most is a dispute test verifying that sector snapshotting prevents valid posts from being disputed when sector set changes.

File: actors/test/replica_update_test.go [hunk 1]
```diff
@@ -123,35 +148,244 @@ func TestReplicaUpdateSuccess(t *testing.T) {
...
+	deadlineInfo, partitionIndex, v = vm.AdvanceTillProvingDeadline(t, v, minerAddrs.IDAddress, sectorNumber)
+
+	vm.SubmitPoSt(t, v, minerAddrs.IDAddress, worker, deadlineInfo, partitionIndex)
+
+	require.True(t, vm.CheckSectorActive(t, v, minerAddrs.IDAddress, deadlineIndex, partitionIndex, sectorNumber))
+	require.False(t, vm.CheckSectorFaulty(t, v, minerAddrs.IDAddress, deadlineIndex, partitionIndex, sectorNumber))
+	minerPower = vm.MinerPower(t, v, minerAddrs.IDAddress)
+	require.Equal(t, uint64(ss), minerPower.Raw.Uint64())
+
+	// TODO: miss a post, lose power
```
[the remark this answers is not in the document]
Comment (arajasek) [reply actors/test/replica_update_test.go:190]: Yeah, they need to get added -- I like the idea of adding them to this test (to have one jumbo start-to-finish full-lifecycle test, but it's not getting into this PR / isn't the most interesting test left to write.

[the remark this answers is not in the document]
Comment (arajasek) [reply actors/test/replica_update_test.go:190]: Removing TODOs and adding to #1513 for now

File: actors/test/replica_update_test.go [hunk 2]
```diff
@@ -123,35 +148,244 @@ func TestReplicaUpdateSuccess(t *testing.T) {
...
+	vm.SubmitPoSt(t, v, minerAddrs.IDAddress, worker, deadlineInfo, partitionIndex)
+
+	require.True(t, vm.CheckSectorActive(t, v, minerAddrs.IDAddress, deadlineIndex, partitionIndex, sectorNumber))
+	require.False(t, vm.CheckSectorFaulty(t, v, minerAddrs.IDAddress, deadlineIndex, partitionIndex, sectorNumber))
+	minerPower = vm.MinerPower(t, v, minerAddrs.IDAddress)
+	require.Equal(t, uint64(ss), minerPower.Raw.Uint64())
+
+}
+
+func TestUnhealthySectorFailure(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (arajasek) [reply actors/test/replica_update_test.go:260]: Added to #1513

File: actors/test/replica_update_test.go [hunk 3]
```diff
@@ -123,35 +148,244 @@ func TestReplicaUpdateSuccess(t *testing.T) {
...
+		NewSealedSectorCID: tutil.MakeCID("replica", &miner.SealedCIDPrefix),
+		Deals:              dealIDs,
+		UpdateProofType:    abi.RegisteredUpdateProof_StackedDrg32GiBV1,
+	}
+
+	vm.ApplyOk(t, v, addrs[0], minerAddrs.RobustAddress, big.Zero(),
+		builtin.MethodsMiner.ProveReplicaUpdates,
+		&miner.ProveReplicaUpdatesParams{Updates: []miner.ReplicaUpdate{replicaUpdate}})
+
+	// immediately miss post, lose power, become faulty
```
[the remark this answers is not in the document]
Comment (arajasek) [reply actors/test/replica_update_test.go:237]: @ZenGround0 I wanted to cover the case of going faulty immediately after a Snap (in the prev test we PoSt the updated sector once before getting into Faulty Business).

You may be right that it's not super different, I thought there was a meaningful distinction between `Snap -> Faulty` and 'Snap -> ProvenSnap -> Faulty`

[the remark this answers is not in the document]
Comment (arajasek) [reply actors/test/replica_update_test.go:237]: Leaving in now and optimistically merging cuz why not, but we can revisit

**Pull request merged**